x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractRuntimeValues.java

```diff
@@ -0,0 +1,175 @@
+                if (lastDocBase == -1) {
+                    // Now that we're started future iterations can't share these values.
+                    unstarted = null;
+                }
+                lastDocBase = ctx.docBase;
+                IntConsumer leafLoader = newLeafLoader(ctx);
+                docId = -1;
+                maxDoc = ctx.reader().maxDoc();
+                lastLeafCursor = new IntConsumer() {
+                    @Override
+                    public void accept(int targetDocId) {
+                        if (docId == targetDocId) {
+                            return;
+                        }
+                        docId = targetDocId;
+                        count = 0;
+                        leafLoader.accept(targetDocId);
+                        if (sort) {
+                            sort();
+                        }
+                    }
+                };
+            }
+            return lastLeafCursor;
+        }
+
+        protected final void alwaysSortResults() {
+            sort = true;
+        }
+
+        protected final int docId() {
+            return docId;
+        }
+
+        protected final int maxDoc() {
+            return maxDoc;
+        }
+
+        protected abstract IntConsumer newLeafLoader(LeafReaderContext ctx) throws IOException;
+
+        protected abstract void sort();
+
+        protected abstract class AbstractRuntimeQuery extends Query {
+            protected final String fieldName;
+
+            protected AbstractRuntimeQuery(String fieldName) {
+                this.fieldName = fieldName;
+            }
+
+            @Override
+            public final Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+                return new ConstantScoreWeight(this, boost) {
+                    @Override
+                    public boolean isCacheable(LeafReaderContext ctx) {
+                        return false; // scripts aren't really cacheable at this point
+                    }
+
+                    @Override
+                    public Scorer scorer(LeafReaderContext ctx) throws IOException {
+                        IntConsumer leafCursor = leafCursor(ctx);
+                        DocIdSetIterator approximation = DocIdSetIterator.all(ctx.reader().maxDoc());
+                        TwoPhaseIterator twoPhase = new TwoPhaseIterator(approximation) {
+                            @Override
+                            public boolean matches() throws IOException {
+                                leafCursor.accept(approximation.docID());
+                                return AbstractRuntimeQuery.this.matches();
+                            }
+
+                            @Override
+                            public float matchCost() {
+                                // TODO we have no idea what this should be and no real way to get one
+                                return 1000f;
```

Comment:
I think the cost can be for now the approximation.cost() , it is proportional to maxDocs if I remember correctly